src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -2,143 +2,180 @@
-            if (projectFilePath.StartsWith(currentDirString))
+            // Open project instance to see if it is a valid project
+            ProjectRootElement projectRootElement = ProjectRootElement.Open(fullPath);
+            SolutionProjectModel project;
+            try
             {
-                projectFilePath = projectFilePath.Substring(currentDirString.Length);
+                project = solution.AddProject(pathRelativeToSolutionFile, null, solutionFolder);
             }
-
-            var projectDirectoryPath = TrimProject(projectFilePath);
-            if (string.IsNullOrEmpty(projectDirectoryPath))
-                return solutionFolders;
-
-            var solutionFoldersPath = TrimProjectDirectory(projectDirectoryPath);
-            if (string.IsNullOrEmpty(solutionFoldersPath))
-                return solutionFolders;
-
-            solutionFolders.AddRange(solutionFoldersPath.Split(Path.DirectorySeparatorChar));
-
-            return solutionFolders;
-        }
-
-        private IList<string> DetermineSolutionFolder(SlnFile slnFile, string fullProjectPath)
-        {
-            if (_inRoot)
+            catch (SolutionArgumentException ex) when (ex.ParamName == "projectTypeName")
             {
-                // The user requested all projects go to the root folder
-                return null;
+                // If guid is not identified by vs-solutionpersistence, check in project element itself
+                var guid = projectRootElement.GetProjectTypeGuid();
+                if (string.IsNullOrEmpty(guid))
+                {
+                    Reporter.Error.WriteLine(CommonLocalizableStrings.UnsupportedProjectType, fullPath);
+                    return;
+                }
+                project = solution.AddProject(pathRelativeToSolutionFile, guid, solutionFolder);
             }
-
-            if (_relativeRootSolutionFolders != null)
+            // Generate intermediate solution folders
+            if (solutionFolder is null && !_inRoot)
             {
-                // The user has specified an explicit root
-                return _relativeRootSolutionFolders;
+                var relativePathDirectory = Path.GetDirectoryName(pathRelativeToSolutionFile);
+                if (!string.IsNullOrEmpty(relativePathDirectory))
+                {
+                    SolutionFolderModel relativeSolutionFolder = solution.AddFolder(GetSolutionFolderPathWithForwardSlashes(relativePathDirectory));
+                    project.MoveToFolder(relativeSolutionFolder);
+                    // Avoid duplicate folder/project names
+                    if (project.Parent is not null && project.Parent.ActualDisplayName == project.ActualDisplayName)
+                    {
+                        solution.RemoveFolder(project.Parent);
+                    }
+                }                
             }
+            // Add settings based on existing project instance
```

Comment:
Note to self: We dont need to specify the global properties since the sln doesn't manage them and it doesn't seem to impact anything.